### ISSUE TITLE
Using SignalR Client version 5.0.0 in Unity3d for mobile application

### ISSUE THREAD

**Issue by bouloss** [dotnet/aspnetcore#27679]
As my title says, I'm trying to use my mobile application (for Android and iOS) as a SignalR client build with Unity3d version 2019.2.9f1.

I began by downloading the latest version of the AspNetCore SignalR client from nuget and extracting the dll. I placed this in my Plugins folder, and repeated the process for each of its dependencies, getting the latest version of each from nuget. 

I included a link.xml file to preserve each of the assemblies in that same folder:

 ```html
<linker>
  <assembly fullname="Microsoft.AspNetCore.Connections.Abstractions" preserve="all"/>
  <assembly fullname="Microsoft.AspNetCore.Http.Connections.Client" preserve="all"/>
  <assembly fullname="Microsoft.AspNetCore.Http.Connections.Common" preserve="all"/>
  <assembly fullname="Microsoft.AspNetCore.Http.Features" preserve="all"/>
  <assembly fullname="Microsoft.AspNetCore.SignalR.Client.Core" preserve="all"/>
  <assembly fullname="Microsoft.AspNetCore.SignalR.Client" preserve="all"/>
  <assembly fullname="Microsoft.AspNetCore.SignalR.Common" preserve="all"/>
  <assembly fullname="Microsoft.AspNetCore.SignalR.Protocols.Json" preserve="all"/>
  <assembly fullname="Microsoft.Extensions.DependencyInjection.Abstractions" preserve="all"/>
  <assembly fullname="Microsoft.Extensions.DependencyInjection" preserve="all"/>
  <assembly fullname="Microsoft.Extensions.Logging.Abstractions" preserve="all"/>
  <assembly fullname="Microsoft.Extensions.Logging" preserve="all"/>
  <assembly fullname="Microsoft.Extensions.Options" preserve="all"/>
  <assembly fullname="Microsoft.Extensions.Primitives" preserve="all"/>
  <assembly fullname="System.Buffers" preserve="all"/>
  <assembly fullname="System.ComponentModel.Annotations" preserve="all"/>
  <assembly fullname="Microsoft.Bcl.AsyncInterfaces" preserve="all"/>
  <assembly fullname="System.Diagnostics.DiagnosticSource" preserve="all"/>
  <assembly fullname="System.IO.Pipelines" preserve="all"/>
  <assembly fullname="System.Memory" preserve="all"/>
  <assembly fullname="System.Text.Encodings.Web" preserve="all"/>
  <assembly fullname="System.Runtime.CompilerServices.Unsafe" preserve="all"/>
  <assembly fullname="System.Text.Json" preserve="all"/>
  <assembly fullname="System.Threading.Channels" preserve="all"/>
  <assembly fullname="System.Threading.Tasks.Extensions" preserve="all"/>
</linker>
```

And my client code:

```c#
using Microsoft.AspNetCore.SignalR.Client;
using System;
using UnityEngine;

public class SignalRManager : MonoBehaviour
{
    public static HubConnection myConnection;

    public void StartSignalConnection()
    {
        try
        {
            Debug.Log("Before building");
            myConnection = new HubConnectionBuilder().WithUrl("ServerURL").Build();
            Debug.Log("After building");

            myConnection.On<string>("MessageClient", message =>
            {
                Debug.Log("From signalR: " + message);
            });

            myConnection.StartAsync();
        }
        catch (Exception ex) { Debug.Log("My error caught here: " + ex.ToString()); }
    }

    public void SendMessage()
    {
        if (myConnection.State != HubConnectionState.Connected) { Debug.Log("Not connected yet"); return; }

        myConnection.InvokeAsync("MessageClient", "Test message");
    }
}
```

In the editor, all of the above works perfectly. I press a button in my scene and it starts the connection, and I press another button and a message is sent. Once I build to my Android device and try to start the connection, I get the following:

```
2020/11/10 12:02:06.793 17686 17728 Info Unity Before building
2020/11/10 12:02:06.793 17686 17728 Info Unity  
2020/11/10 12:02:06.793 17686 17728 Info Unity (Filename: ./Runtime/Export/Debug/Debug.bindings.h Line: 35)
2020/11/10 12:02:06.793 17686 17728 Info Unity 
2020/11/10 12:02:06.885 17686 17728 Error Unity Unable to find advapi32
2020/11/10 12:02:06.886 17686 17728 Warn Unity Plugins: Couldn't open advapi32, error: dlopen failed: library "advapi32" not found
2020/11/10 12:02:06.886 17686 17728 Warn Unity  
2020/11/10 12:02:06.886 17686 17728 Warn Unity (Filename: ./Runtime/Misc/Plugins.cpp Line: 282)
2020/11/10 12:02:06.886 17686 17728 Warn Unity 
2020/11/10 12:02:06.892 17686 17728 Warn Unity Plugins: Couldn't open advapi32, error: dlopen failed: library "advapi32" not found
2020/11/10 12:02:06.892 17686 17728 Warn Unity  
2020/11/10 12:02:06.892 17686 17728 Warn Unity (Filename: ./Runtime/Misc/Plugins.cpp Line: 282)
2020/11/10 12:02:06.892 17686 17728 Warn Unity 
2020/11/10 12:02:07.066 17686 17728 Info Unity After building
2020/11/10 12:02:07.066 17686 17728 Info Unity  
2020/11/10 12:02:07.066 17686 17728 Info Unity (Filename: ./Runtime/Export/Debug/Debug.bindings.h Line: 35)
2020/11/10 12:02:07.066 17686 17728 Info Unity 
```
I looked into what advapi32 is and what it's meant for, but I'm not sure how to work around needing it, as I'm not sure I can just pop it into the folder if I can get the file. Should I be taking a different approach here to get the client working in my app? If there's any way to get a more detailed stack trace then I'm more than willing to do so. If this isn't the right place to be directing my question, then please let me know that as well.

**Comment by BrennanConroy** [comment]
If you can figure out how to get a stack trace and it shows an error inside SignalR then we might be able to look into it. Otherwise, this should probably be filed on the Unity forum.

**Comment by bouloss** [comment]
Apologies for the delay on this, as I've been trying to get more detail.

```
Warn Unity Plugins: Couldn't open advapi32, error: dlopen failed: library "advapi32" not found
Warn Unity Microsoft.Win32.ManifestEtw:EventRegister(Guid&, EtwEnableCallback, Void*, Int64&)
Warn Unity System.Diagnostics.Tracing.EventProvider:Register(Guid)
Warn Unity System.Diagnostics.Tracing.EventSource:Initialize(Guid, String, String[])
Warn Unity Microsoft.Extensions.DependencyInjection.DependencyInjectionEventSource:.cctor()
Warn Unity Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine:CreateServiceAccessor(Type)
Warn Unity System.Func`2:Invoke(T)
Warn Unity System.Collections.Concurrent.ConcurrentDictionary`2:GetOrAdd(TKey, Func`2)
Warn Unity Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine:GetService(Type, ServiceProviderEngineScope)
Warn Unity Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions:GetService(IServiceProvider)
Warn Unity Microsoft.AspNetCore.SignalR.Client.HubConnectionBuilder:Build()
Warn Unity SignalR:StartSignalRConnection()
Warn Unity UnityEngine.Events.UnityAction:Invoke()
Warn Unity UnityEngine.Events.UnityEvent:Invoke()
Warn Unity UnityEngine.EventSystems.EventFunction

Warn Unity Plugins: Couldn't open advapi32, error: dlopen failed: library "advapi32" not found
Warn Unity Microsoft.Win32.ManifestEtw:EventRegister(Guid&, EtwEnableCallback, Void*, Int64&)
Warn Unity System.Diagnostics.Tracing.EventProvider:Register(Guid)
Warn Unity System.Diagnostics.Tracing.EventSource:Initialize(Guid, String, String[])
Warn Unity System.Buffers.ArrayPoolEventSource:.cctor()
Warn Unity System.Buffers.DefaultArrayPool`1:Rent(Int32)
Warn Unity System.Text.Json.JsonEncodedText:TranscodeAndEncode(ReadOnlySpan`1, JavaScriptEncoder)
Warn Unity System.Text.Json.JsonEncodedText:Encode(ReadOnlySpan`1, JavaScriptEncoder)
Warn Unity System.Text.Json.JsonEncodedText:Encode(String, JavaScriptEncoder)
Warn Unity Microsoft.AspNetCore.SignalR.Protocol.JsonHubProtocol:.cctor()
Warn Unity System.Reflection.MonoCMethod:InternalInvoke(Object, Object[])
Warn Unity Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver:VisitConstructor(ConstructorCallSite, RuntimeResolverContext)
Warn Unity Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver:VisitCache(ServiceCallSite, RuntimeResolverContext, ServiceP

Warn Unity Plugins: Couldn't open advapi32, error: dlopen failed: library "advapi32" not found
Warn Unity Microsoft.Win32.ManifestEtw:EventRegister(Guid&, EtwEnableCallback, Void*, Int64&)
Warn Unity System.Diagnostics.Tracing.EventProvider:Register(Guid)
Warn Unity System.Diagnostics.Tracing.EventSource:Initialize(Guid, String, String[])
Warn Unity System.Net.NetEventSource:.cctor()
Warn Unity System.Net.WebSockets.ClientWebSocket:.ctor()
Warn Unity Microsoft.AspNetCore.Http.Connections.Client.HttpConnection:IsWebSocketsSupported()
Warn Unity Microsoft.AspNetCore.Http.Connections.Client.Log:ErrorWithNegotiation(ILogger, Uri, Exception)
Warn Unity System.Threading.ContextCallback:Invoke(Object)
Warn Unity System.Threading.ExecutionContext:RunInternal(ExecutionContext, ContextCallback, Object, Boolean)
Warn Unity System.Runtime.CompilerServices.MoveNextRunner:Run()
Warn Unity System.Action:Invoke()
Warn Unity System.Threading.ContextCallback:Invoke(Object)
Warn Unity System.Threading.Tasks.AwaitTaskContinuation:RunCallback(ContextCallback, Object, Task&)
Warn Unity System.Threading.Tasks.Task:FinishContinuations()
Warn Unity System.Threading.Tasks.Task`1:TrySetResult(TResult)
```

I was able to get the above logs. These all occurred in the following line: 

`
myConnection = new HubConnectionBuilder().WithUrl("ServerURL").Build();
`

I can see that the issue here is still being unable to find advapi32.dll. I looked to add this to my project, but while it may resolve the issue of the dll not being found, I doubt that would resolve the whole issue since advapi32 is used for security and registry calls in the base Windows 32 API. Is there any way that I would be able to get around this dependency?

**Comment by BrennanConroy** [comment]
It looks like anything that initializes an EventSource will fail with this error. You might want to ask on the Unity forums if this is expected.

As a workaround you can try setting the "System.Diagnostics.Tracing.EventSource.IsSupported" app context switch to false at the beginning of your app. https://docs.microsoft.com/dotnet/api/system.appcontext.setswitch?view=net-5.0

**Comment by bouloss** [comment]
I gave the context switch a shot, but unfortunately the result was the same. I ended up getting around this problem by utilizing the [Best HTTP/2 Asset](https://assetstore.unity.com/packages/tools/network/best-http-2-155981) from the Unity Asset store. This implemented SignalR in a way that worked in Unity for mobile builds. I'll go ahead and close out this issue now, as the technologies seem to be working as intended on the AspNetCore end of things.
